File: java/Arrays/Day23/MergeTwoArrayWES.java
package Day23;
import java.util.*;

public class MergeTwoArrayWES {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter the size of arr1: ");
        int n = sc.nextInt();

        System.out.print("Enter the size of arr2: ");
        int m = sc.nextInt();

        System.out.print("Enter " + n + " " + "numbers from arr1: ");
        int[] arr1 = new int[n];
        for (int i = 0; i < n; i++) {
            arr1[i] = sc.nextInt();
        }

        System.out.print("Enter " + m + " " + "numbers from arr2: ");
        int[] arr2 = new int[m];
        for (int j = 0; j < m; j++) {
            arr2[j] = sc.nextInt();
        }




        // -------- BRUTE APPROACH FOR MERGE 2 SORTED ARRAY WITH USING EXTRA SPACE(WES)  ---------
        // int[] arr3 = new int[n + m];
        // int left = 0, right = 0;
        // int index = 0;

        // while (left < n && right < m) {
        //     if (arr1[left] <= arr2[right]) {
        //         arr3[index] = arr1[left];
        //         left++; index++;
        //     } else {
        //         arr3[index] = arr2[right];
        //         right++; index++;
        //     }
        // }

        // while (left < n) {
        //     arr3[index++] = arr1[left++];
        // }

        // while (right < m) {
        //     arr3[index++] = arr2[right++];
        // }

        // for (int i = 0; i < n + m; i++) {
        //     if (i < n) {
        //         arr1[i] = arr3[i];
        //     } else {
        //         arr2[i - n] = arr3[i];
        //     }
        // }







        // -------- FIRST OPTIMAL APPROACH FOR MERGE 2 SORTED ARRAY WITHOUT USING EXTRA SPACE(WES) ---------
        // int left = n - 1, right = 0;

        // while (left >= 0 && right < m) {
        //     if (arr1[left] > arr2[right]) {
        //         swap(arr1, arr2, left, right);
        //         left--; right++;
        //     } else {
        //         break;
        //     }
        // }

        // Arrays.sort(arr1);
        // Arrays.sort(arr2);






        // -------- SECOND OPTIMAL APPROACH FOR MERGE 2 SORTED ARRAY WITHOUT USING EXTRA SPACE(WES) ---------
        // // Usign gap method (shell sort)
        int len = n + m;
        int gap = (len / 2) + (len % 2);

        while (gap > 0) {
            int left = 0;
            int right = left + gap;

            while (right < len) {
                // arr1 and arr2...
                if (left < n && right >= n) {
                    if (arr1[left] > arr2[right - n]) {
                        swap(arr1, arr2, left, right - n);
                    }
                }

                // arr2 and arr2...
                else if (left >= n) {
                    if (arr2[left - n] > arr2[right - n]) {
                        swap(arr2, arr2, left - n, right - n);
                    }
                }

                // arr1 and arr1...
                else {
                    if (arr1[left] > arr1[right]) {
                        swap(arr1, arr1, left, right);
                    }
                }

                left++; right++;
            }

            if (gap == 1) break;
            gap = (gap / 2) + (gap % 2);
        }







        // Print results
        System.out.println("Merged arr1: " + Arrays.toString(arr1));
        System.out.println("Merged arr2: " + Arrays.toString(arr2));

        sc.close();
    }

    
    // swap function...
    public static void swap(int[] arr1, int[] arr2, int left, int right) {
        int temp = arr1[left];
        arr1[left] = arr2[right];
        arr2[right] = temp;
    }
}

File: python/Arrays/day23/mergeTwoArrayWES.py
n = int(input("Enter the size of arr1: "))
m = int(input("Enter the size of arr2: "))

arr1 = list(map(int, input(f"Enter {n} numbers from arr1: ").split()))
arr2 = list(map(int, input(f"Enter {m} numbers from arr2: ").split()))



# -------- BRUTE FORCE APPROACH ---------
# arr3 = [0] * (n + m)
# left, right, index = 0, 0, 0

# while left < n and right < m:
#     if arr1[left] <= arr2[right]:
#         arr3[index] = arr1[left]
#         left += 1
#         index += 1
#     else:
#         arr3[index] = arr2[right]
#         right += 1
#         index += 1

# while left < n:
#     arr3[index] = arr1[left]
#     left += 1
#     index += 1

# while right < m:
#     arr3[index] = arr2[right]
#     right += 1
#     index += 1


# for i in range(0, n + m):
#     if i < n:
#         arr1[i] = arr3[i]
#     else:
#         arr2[i - n] = arr3[i]







# -------- FIRST OPTIMAL APPROACH ---------
# left = n - 1
# right = 0

# while left >= 0 and right < m:
#     if arr1[left] > arr2[right]:
#         arr1[left], arr2[right] = arr2[right], arr1[left]
#         left -= 1
#         right += 1
#     else:
#         break

# arr1.sort()
# arr2.sort()






# -------- SECOND OPTIMAL APPROACH ---------
len_ = (n + m)
gap = (len_ // 2) + (len_ % 2)

while gap > 0:
    left = 0
    right = left + gap

    while right < len_:
        # arr1 and arr2
        if left < n and right >= n:
            if arr1[left] > arr2[right - n]:
                arr1[left], arr2[right - n] = arr2[right - n], arr1[left]
        
        # arr2 and arr2
        elif left >= n:
            if arr2[left - n] > arr2[right - n]:
                arr2[left - n], arr2[right - n] = arr2[right - n], arr2[left - n]
        
        # arr1 and arr2
        else:
            if arr1[left] > arr1[right]:
                arr1[left], arr1[right] = arr1[right], arr1[left]
        
        left += 1
        right += 1

    if gap == 1:
        break

    gap = (gap // 2) + (gap % 2)









# Print results
print("Merged arr1:", arr1)
print("Merged arr2:", arr2)